compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -83,3 +94,187 @@ fn search_for_metadata<'a>(
         .data()
         .map_err(|e| format!("failed to read {} section in '{}': {}", section, path.display(), e))
 }
+
+fn create_object_file(sess: &Session) -> Option<write::Object> {
+    let endianness = match sess.target.options.endian {
+        Endian::Little => Endianness::Little,
+        Endian::Big => Endianness::Big,
+    };
+    let architecture = match &sess.target.arch[..] {
+        "arm" => Architecture::Arm,
+        "aarch64" => Architecture::Aarch64,
+        "x86" => Architecture::I386,
+        "s390x" => Architecture::S390x,
+        "mips" => Architecture::Mips,
+        "mips64" => Architecture::Mips64,
+        "x86_64" => {
+            if sess.target.pointer_width == 32 {
+                Architecture::X86_64_X32
+            } else {
+                Architecture::X86_64
+            }
+        }
+        "powerpc" => Architecture::PowerPc,
+        "powerpc64" => Architecture::PowerPc64,
+        "riscv32" => Architecture::Riscv32,
+        "riscv64" => Architecture::Riscv64,
+        "sparc64" => Architecture::Sparc64,
+        // Unsupported architecture.
+        _ => return None,
+    };
+    let binary_format = if sess.target.is_like_osx {
+        BinaryFormat::MachO
+    } else if sess.target.is_like_windows {
+        BinaryFormat::Coff
+    } else {
+        BinaryFormat::Elf
+    };
+
+    let mut file = write::Object::new(binary_format, architecture, endianness);
```

Comment:
This should probably get a `file.add_file_symbol(...)` call. In cg_clif I used the cgu name as argument.